include/envoy/event/timer.h

```diff
@@ -30,7 +30,12 @@ class Timer {
   /**
    * Enable a pending timeout. If a timeout is already pending, it will be reset to the new timeout.
    */
-  virtual void enableTimer(const std::chrono::milliseconds& d) PURE;
+  virtual void enableTimerInUs(const std::chrono::microseconds& d) PURE;
+
+  virtual void enableTimer(const std::chrono::milliseconds& d) {
```

Comment:
I see what you are trying to do but I have a slightly different suggestion. See test/test_common/test_time_system.h and the definitions of sleep() for the pattern. It references TimeSystem::Duration which is defined in include/envoy/event/timer.h, and does not tie the caller to any particular unit.

That way it can be up to the caller whether to specify milliseconds or microseconds or whatever.

Reply:
Are you suggesting adding another member function like `void enableTimerAfterDuration(Duration d)`?

Reply:
Yes, but I was suggesting naming it enableTimer(Duration d) as an inline template method, to avoid having to change any existing callsites.

Look at the pattern in test_time_system.h.

Reply:
Sorry, I misunderstood your suggestion. Please ignore my previous question.
You mean change Timer interface to be something as below right?

  virtual void enableTimer(const MonotonicTime::duration& d) PURE;
  template<class D> void enableTimer(const D& duration) {
    enableTimer(std::chrono::duration_cast<MonotonicTime::duration>(duration));
  }

But doing so still requires changes to all EXPECT_CALL(timer, enableTime(some_duration)). I don't have strong preference to this template approach v.s. adding enableTimerInUs(). If the refactor takes some large scope change, I would prefer the former one actually.

Reply:
I think changing all EXPECT_CALLs is fine, but I'm not a big fan of using the same name for the virtual and template functions, it just feel very wired.

Long term, I'd suggest we use a duration type that does not imply the unit of time, like absl::Duration, for the enableTimer interface, but it's out of the scope of this PR.

Reply:
What's the conclusion then? 
The purpose of adding a templated method with same name is to minimize the changes on the call site. But if EXPECT_CALLs need change anyway, this approach is not more preferable than current approach.

Reply:
Yeah the mocks are annoying and there are >500 of them.

I'd still like to wind up with a crisp single entry-point for users, which I think is easiest to do with a templated method as suggested.

The goal isn't necessarily to minimize changes; the goal is to have a clean interface that will last for a long time across many use-cases.

So my vote would be to have a protected virtual helper method and change the mocks to override that. I know this is annoying because there are >500 of them but it should be an easy script to fix them.

That should be a separate PR that has a trivial large rename in it but no new functionality.

Reply:
And ultimately the mocking of enableTimer is usually unnecessary given simulated time IMO, but I wouldn't want to take on that debt to clean those up :)

Reply:
Can I add `virtual void enableTimer(const MonotonicTime::duration& d)` as protected member and keep the old `virtual void enableTimer(const std::chrono::milliseconds& d)` as public? I.e.
class Timer {
public:
   virtual void enableTimer(const std::chrono::milliseconds& d) {
     enableTimer(std::chrono::duration_cast<MonotonicTime::duration>(d));
   } 
  template void enableTimer(const D& duration) {
    enableTimer(std::chrono::duration_cast<MonotonicTime::duration>(duration));
  }
protected:
   virtual void enableTimer(const MonotonicTime::duration& d) PURE;
}

So the templated enableTimer() can call the private member but the call sites can call with any unit because of the template and MOCK_METHOD1(enableTimer, void(const std::chrono::milliseconds&)) still works.

Reply:
that sounds ideal. Actually that was my initial suggestion (the pattern I referenced above), and was assuming it didn't work due to mocking. Does it actually work?

Reply:
The difference is that we still keep the old enableTimer(const std::chrono::milliseconds&) there. Is that ok?

Reply:
Yeah I think that'd be OK if it works.  I've had difficulties in the past overriding one of two overloaded virtual methods, which I think is what we would happen with mocks. It's worth trying.

Reply:
Anyway I wind to wind up in a state where there's one enableTimer entry that takes a Duration rather than having variants for different time units. It might be simplest to start a separate PR that makes that change, including some boring changes to mocks.

Reply:
+1 to single enableTimer taking a Duration. It seems to have higher resolution than milliseconds and low resolution duration types(e.g. std::chrono::seconds) can implicitly convert to it.

Reply:
Agree to do it in separate PR. Filed https://github.com/envoyproxy/envoy/issues/6588 for tracking.

I talked to Bin and Ian offline, having milliseconds granularity is fine for now. In QUICHE code, QuicAlarm has always been scheduled with (Now + some milliseconds) as RTT is usually in ms granularity. The only exception comes when it is scheduled with Now in which case casting now to milliseconds doesn't change the expected behavior. So I will use existing enbableTimer() for now.